### ISSUE TITLE
Vibration no longer working with new firmware

### ISSUE THREAD

**Issue by eladnava** [betomaluje/Mi-Band#4]
Hi @betomaluje,
First of all, thanks for this awesome Android library!

I'm thinking of integrating it with my [RedAlert](https://github.com/eladnava/redalert-android) app, so that deaf people will know there is a rocket alert in their area (via vibration mostly).

I tried the `MiBandExample` project, and it works pretty nicely (everything except for the vibration, which doesn't really do anything for some reason) -- is this something you are aware of?

My bet is that they changed something in the new firmware.

Thanks!

**Comment by eladnava** [comment]
Also, I just tried the following and none of them worked:

```
    public static final int ACTION_VIBRATE_WITH_LED = 3;
    public static final int ACTION_VIBRATE_UNTIL_CALL_STOP = 4;
    public static final int ACTION_VIBRATE_WITHOUT_LED = 5;
    public static final int ACTION_VIBRATE_CUSTOM = 6;
```

**Comment by eladnava** [comment]
This may help:
https://github.com/lwis/miband-notifier/issues/33

**Comment by betomaluje** [comment]
I sniffed the Bluetooth with the band and sent the same commands using the nRF Master Control Panel app and using the same commands worked like a charm.

Try using the latest version of the library and use the ``miBand.startVibration(VibrationMode.VIBRATION_WITHOUT_LED);`` command to test it.

Let me know if the problem persists.

**Comment by eladnava** [comment]
@betomaluje tried it, unfortunately, no vibration -- is it possible that the Mi Fit app updated my band's firmware recently (and your firmware hasn't been updated yet because you don't have that app installed)?

**Comment by betomaluje** [comment]
I have tweaked version of Mi Fit installed. My firmware version is 1.0.10.11 . Is this the same as yours?

**Comment by eladnava** [comment]
@betomaluje nope, I've got the official Mi Fit app (and my firmware version is now 1.0.**11.6**).

By the way, thanks a lot for helping with this!

**Comment by eladnava** [comment]
@betomaluje after some excruciatingly-painful smali debugging of [Mi Band Notifier v2](https://play.google.com/store/apps/details?id=com.mc.miband1), I was able to extract the new characteristic `2A06` and the new vibrate (without light) command: `0x4`.

However, the `2A06` characteristic isn't available in the `mAvailableCharacteristics` so `getCharacteristic()` fails. Any idea how to fix that?

**Comment by betomaluje** [comment]
@eladnava thank you for your research! Unfortunately I wasn't able to send this commands to the band. I tried reading the characteristic inside ``BTConnectionManager`` and using a few BT apps such as nRF Master Control Panel but none of them could found the ``2A06`` characteristic. I tried sending them an nothing happened. I used several combinations changing the ``Profile`` and the ``Protocol`` clases but nothing happened.

``2A06`` is used inside the ``Profile`` as the ``UUID_CHAR_ALERT_LEVEL``. Also tried changing the Write command on the ``startVibration`` method inside ``MiBand`` to this one but no results.

There must be something that I'm doing wrong. You can play with those values and let me know if anything works. I'm also going to research why this isn't working.

Hope this helps you someway.

**Comment by eladnava** [comment]
@betomaluje did you update your band's firmware to 1.0.11.6?

I did go ahead and change the values for the VIBRATION_NO_LED (`0x4` via `2A06` characteristic), but then I discovered that this characteristic is not available for the Mi Band device.

I'll try to use the [nRF App](https://play.google.com/store/apps/details?id=no.nordicsemi.android.mcp&hl=en) to sniff out the Mi Band Notify v2, just in case.

**Comment by betomaluje** [comment]
@eladnava yes I did. Sorry I forgot to mention that. Let me know if anything helped you.

For the sniffing part I was following [this tutorial] (https://learn.adafruit.com/reverse-engineering-a-bluetooth-low-energy-light-bulb/sniff-protocol). But I don't have the hardware. Also I found [this guide] (https://www.nowsecure.com/blog/2014/02/07/bluetooth-packet-capture-on-android-4-4/) for doing the same thing but using an Android phone but I haven't had the time to give it a try. In theory, you can listen to each command sent back and forth between the band an the official app, save the log to the SD card and then analyse the data using [Wireshark] (https://www.wireshark.org/#download). If you found the exact command then we can replicate it on the app.

Hope this helps!

**Comment by eladnava** [comment]
@betomaluje thanks!

Just to make sure, are the available Bluetooth characteristics provided by the Mi Band? Because the `2A06` characteristic isn't listed for some reason. Or are they determined by the app?

**Comment by betomaluje** [comment]
@eladnava I really don't remember where did I found those characteristics but where the default characteristics for bluetooth devices. The Xiaomi guys haven't provided the characteristics they use so at this point I was just trying things.

I can't find the ``2A06`` characteristic but my guess is that using the Wireshark approach may help. 

Let me know if you find something

**Comment by eladnava** [comment]
@betomaluje I did it! =)

It turns out that the `0x4` via the `2A06` characteristic is correct -- however, the service was wrong -- instead of using the default (`FEE0`), I was able to find the characteristic within the `1802` service.

Would you like me to submit a pull request updating the library with the new firmware's vibrate codes? Also, do we need to offer backward compatibility for old firmware versions?

**Comment by betomaluje** [comment]
@eladnava You're the man! that's awesome! Please submit that pull request. The compatibility would be great. There must be a way of seeing the current firmware and if it's the newest, use your method. I would investigate if there's a way of getting the firmware (there must be a way, the Mi Fit app does it).

Great work!

**Comment by eladnava** [comment]
@betomaluje I see that you already implemented it yourself. :)

Closed via a8a32209065955bf26b36d0d8d4714594a0855e7.
